### PR TITLE
chore(pack): Use electron-builder for packaging

### LINKED ISSUE #915: Try out electron-builder
See how nice it would be to adapt our setup to using [electron-builder](https://github.com/electron-userland/electron-builder).

### PR DESCRIPTION
**Pull request by lgeiger**
Try out [`electron-builder`](https://github.com/electron-userland/electron-builder): Closes #915 

bildschirmfoto 2016-10-08 um 23 18 23

I don't know how the two `package.json` structure will work with greenkeeper.
- [x] OS X (`.dmg`, `.app`)
- [x] Linux (unpacked, `.AppImage`)
- [x] Windows (unpacked, NSIS installer)
- [x] OS X signing
- [ ] Windows signing

### CODE REVIEW

**Comment by captainsafia** [comment]
> I don't know how the two package.json structure will work with greenkeeper.

It looks like greenkeeper just ignores `package.json`s in sub-directories ([for now](https://github.com/greenkeeperio/greenkeeper/issues/139)) so it'll be up to us to keep `app/package.json` green.

**Comment by lgeiger** [comment]
I'm confident travis will be green with the changes made in #940 (except the version bump).

**Comment by jdetle** [comment]
```
npm WARN nteract No repository field.
npm WARN nteract No license field.
```

I think we lost some stuff in `package.json`? Other than that running `npm run pack` works great on linux.

**Comment by lgeiger** [comment]
@jdetle I'll take a look at the package.json

Just curios: What does `npm run dist` produce on linux?

**Comment by jdetle** [comment]
It produces an `app/` folder containing `build/` `node_modules/` `package.json` and `static/`

**Comment by lgeiger** [comment]
I was hoping it produces packages out of the box.

**Comment by lgeiger** [comment]
And I don't know if this will break npm publish 😕 .

**Comment by rgbkrk** [comment]
> And I don't know if this will break npm publish 

Guess we'll find out. I've only been doing npm publish to bump the version and, for later, for us to handle nteract as a monorepo (#549) where we can use the components in other projects.

Working first is awesome, we can adapt later for the monorepo stuff.

**Comment by lgeiger** [comment]
If two `package.json` files turn out to be a hassle we can always try to do everything manually. Even if it's not recommended.

**Comment by rgbkrk** [comment]
It's ok if we lose greenkeeper, it used to be me running `npm-check -u` and submitting PRs with updates.

**Comment by lgeiger** [comment]
If this is the way we like to go, I'll go through the `electron-builder` docs and fine tune everything.

**Comment by rgbkrk** [comment]
For the signing, I have keys. I'm _slightly_ willing to put them up on Travis, encrypted. I'd rather not, but hey.

**Comment by lgeiger** [comment]
@rgbkrk Publishing from CI is definitely not happening in this PR 😉

**Comment by lgeiger** [comment]
All the tests fail because the node modules can not be found since they are now located in the `app/node_modules` instead of the `node_modules` folder. I just realized it now because travis caches the modules.

~~@rgbkrk Do you know a way to fix this our should I try to use `electron-builder` without the recommended two `package.json` structure?~~

**Comment by lgeiger** [comment]
- OS X:
  - `.dmg` and `.app` will be created
  - file association works automatically
- linux:
  - `.AppImage` and unpacked app will be created
  - I didn't enable `.deb` because the file association didn't work
  - file association works if nteract is selected in the "open with" dialog
- windows:
  - unpacked app and NSIS installer will be created
  - I preferred NSIS over Squirrel because it creates start menu and desktop symlinks
  - file association works if the full path to nteract is selected in the "open with" dialog

**Comment by lgeiger** [comment]
@rgbkrk Could you test the signing on OS X?

**Comment by rgbkrk** [comment]
I'm assuming you cleared the cache to get Travis to work? Looking good.

I'll try the signed builds now.

**Comment by lgeiger** [comment]
> I'm assuming you cleared the cache to get Travis to work?

[`9ca24ac`](https://github.com/nteract/nteract/pull/952/commits/9ca24aca573c8af4dfc01200f1c9d01dd5f19462) and [`4cc12c6`](https://github.com/nteract/nteract/pull/952/commits/4cc12c62c19489e7739425d4d2b8c6e9b960abd0) fixed travis.

> I'll try the signed builds now.

:+1: :tada:

**Comment by rgbkrk** [comment]
```
$ npm i

> undefined postinstall /Users/kyle/code/src/github.com/nteract/nteract
> install-app-deps

Installing app dependencies for arch x64 to /Users/kyle/code/src/github.com/nteract/nteract/app

Error: /Users/kyle/.nvm/versions/node/v6.2.0/bin/node exited with code 1
Error output:
npm ERR! Darwin 15.6.0
npm ERR! argv "/Users/kyle/.nvm/versions/node/v6.2.0/bin/node" "/Users/kyle/.nvm/versions/node/v6.2.0/lib/node_modules/npm/bin/npm-cli.js" "install" "--production" "--cache-min" "999999999" "--build-from-source"
npm ERR! node v6.2.0
npm ERR! npm  v3.8.9

npm ERR! No compatible version found: github@^5.0.0
npm ERR! Valid install targets:
npm ERR! 4.0.0, 3.1.1, 3.1.0, 3.0.0, 2.6.0, 2.5.2, 2.5.1, 2.4.1, 2.4.0, 2.3.0, 2.2.0, 2.1.0, 2.0.1, 2.0.0, 1.4.0, 1.3.0, 1.2.1, 1.2.0, 1.1.2, 1.1.1, 1.1.0, 0.2.4, 0.2.3, 0.2.2, 0.2.1, 0.2.0, 0.1.16, 0.1.15, 0.1.14, 0.1.13, 0.1.12, 0.1.11, 0.1.10, 0.1.9, 0.1.8, 0.1.7, 0.1.6, 0.1.5, 0.1.4, 0.1.3, 0.1.2, 0.1.1, 0.1.0, 0.0.8, 0.0.7, 0.0.6, 0.0.5, 0.0.4, 0.0.3, 0.0.2, 0.0.1
npm ERR!
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

```

**Comment by rgbkrk** [comment]
That's weird -- it's what we use on master and certainly the most recent release: https://www.npmjs.com/package/github

**Comment by rgbkrk** [comment]
I just set it to `github@4.0.0` locally for now so I can progress further.

**Comment by rgbkrk** [comment]
```
Packaging for platform darwin x64 using electron 1.4.3 to dist/mac

node_modules/zmq-prebuilt is not packed into asar archive - contains executable code
Signing app (identity: Developer ID Application: Kyle Kelley (8E54RDNN6Z))
```

**Comment by rgbkrk** [comment]
Worked quite well, I'll post a dmg to slack.

**Comment by rgbkrk** [comment]
Looks like the dmg is working quite well!

**Comment by rgbkrk** [comment]
Before we merge, I'd like @jdetle and @captainsafia to try out this branch, see what the developer workflow is like. Trying out `npm run dist` would be good too.

**Comment by jdetle** [comment]
Unfortunately, I'm running into some problems on Linux. Don't have time this morning to debug, but here is the error: 

```
Unhandled rejection Error: Exit code: ENOENT. spawn icns2png ENOENT
    at /home/jdetlefs/github/nteract/node_modules/electron-builder/src/util/util.ts:102:16
    at exithandler (child_process.js:213:5)
    at ChildProcess.errorhandler (child_process.js:225:5)
    at emitOne (events.js:96:13)
    at ChildProcess.emit (events.js:188:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:213:12)
    at onErrorNT (internal/child_process.js:359:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
From previous event:
    at Object.exec (/home/jdetlefs/github/nteract/node_modules/electron-builder/src/util/util.ts:80:3)
    at LinuxTargetHelper.<anonymous> (/home/jdetlefs/github/nteract/node_modules/electron-builder/src/targets/LinuxTargetHelper.ts:133:32)
From previous event:
    at tsAwaiter (/home/jdetlefs/github/nteract/node_modules/electron-builder/src/util/awaiter.ts:10:47)
    at Object.build (/home/jdetlefs/github/nteract/node_modules/electron-builder/out/builder.js:138:12)
    at Object.<anonymous> (/home/jdetlefs/github/nteract/node_modules/electron-builder/out/build-cli.js:23:11)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.runMain (module.js:590:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3

node_modules/zmq-prebuilt is not packed into asar archive - contains executable code
Error: Exit code: ENOENT. spawn icns2png ENOENT
    at /home/jdetlefs/github/nteract/node_modules/electron-builder/src/util/util.ts:102:16
    at exithandler (child_process.js:213:5)
    at ChildProcess.errorhandler (child_process.js:225:5)
    at emitOne (events.js:96:13)
    at ChildProcess.emit (events.js:188:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:213:12)
    at onErrorNT (internal/child_process.js:359:16)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)
From previous event:
    at Object.exec (/home/jdetlefs/github/nteract/node_modules/electron-builder/src/util/util.ts:80:3)
    at LinuxTargetHelper.<anonymous> (/home/jdetlefs/github/nteract/node_modules/electron-builder/src/targets/LinuxTargetHelper.ts:133:32)
From previous event:
    at tsAwaiter (/home/jdetlefs/github/nteract/node_modules/electron-builder/src/util/awaiter.ts:10:47)
    at Object.build (/home/jdetlefs/github/nteract/node_modules/electron-builder/out/builder.js:138:12)
    at Object.<anonymous> (/home/jdetlefs/github/nteract/node_modules/electron-builder/out/build-cli.js:23:11)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.runMain (module.js:590:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

**Comment by lgeiger** [comment]
@jdetle The [electron-builder instructions](sudo apt-get install --no-install-recommends -y icnsutils graphicsmagick xz-utils) suggest: `sudo apt-get install --no-install-recommends -y icnsutils graphicsmagick xz-utils`

**Comment by jdetle** [comment]
Ah, well then I imagine it will work fine when I do that. Sorry I couldn't take the time to debug.

**Comment by rgbkrk** [comment]
🤔 I wonder if we could run preinstall diagnostics that would, per OS, check to see if the developer has the right system dependencies and then let them know what to do/install.

**Comment by jdetle** [comment]
^ We could add this into the diagnostics script.

**Comment by lgeiger** [comment]
@rgbkrk Preinstall diagnostics would be awesome for zmq-prebuilt. The zmq build errors get piped to `zmq-build.log` because the gyp file didn't work well with stdout. This means the build errors aren't really helpful to debug.

**Comment by lgeiger** [comment]
Let me know if I should change anything in this PR.

For reviewing I recommend looking at every commit separately. The first commit changes the directory structure this makes the overall diff so intimidating.

**Comment by rgbkrk** [comment]
I'll give it another go and review. We should probably rip this band aid off soon.

**Comment by rgbkrk** [comment]
I think the main thing we'll need addressed is an update to the README since the build process has changed. Other than that, I think this is fit to ship.

**Comment by lgeiger** [comment]
@rgbkrk Thanks for taking a look. I'll update the README.

File: README.md [hunk 1]
```diff
@@ -120,22 +120,26 @@ And then opening `docs/index.html` in your favorite browser.
 ### Bump the version
 
 ```
+cd app
 npm version {major|minor|patch}
+cd ..
 git push
 git push --tags
-npm publish
```
Comment (lgeiger) [inline README.md:126]: I'm pretty sure the two `package.json` structure breaks the `npm` release 😕 .

Comment (rgbkrk) [reply README.md:126]: Well we need to focus some efforts on exporting these components for external use soon anyways. I'll have look here in a moment to see how the `npm` release flow can work. We can always create tags manually.

Comment (lgeiger) [reply README.md:126]: I'm sure we could also make it work with a single package.json if we end up needing it.

Comment (rgbkrk) [reply README.md:126]: With the monorepo + lerna stuff, we'll end up having N `package.json`s.

Comment (lgeiger) [reply README.md:126]: I was just stressing the it because I din't know if the npm release is important.

**Comment by rgbkrk** [comment]
:shipit:

**Comment by rgbkrk** [comment]
After pulling I had to:

```
rm -rf build/main/ build/notebook/
```

**Comment by rgbkrk** [comment]
Getting extreme strangeness on startup.

screen shot 2016-10-14 at 2 36 48 pm

screen shot 2016-10-14 at 2 37 16 pm

**Comment by rgbkrk** [comment]
Oh right, I have to do the `npm install` over again because of the new path for `node_modules` in the main app.

**Comment by rgbkrk** [comment]
Something is up with the electron logger though:

```
[14:39:59:0554] [info] args { _: [],
  kernel: 'python3',
  k: 'python3',
  '$0': 'node_modules/electron/dist/Electron.app/Contents/MacOS/Electron .' }
[14:39:59:0561] [warning] electron-log cannot read a name from package.json
[14:39:59:0561] [warning] electron-log.transports.file: Could not set a log file
[14:40:00:0553] [info] launching an empty notebook by default
```

**Comment by lgeiger** [comment]
I recommend running `npm prune` in the root directory to free some disk space.

Maybe we need to add the name to the main package.json again to get rid of the electron-log warning.

**Comment by probonopd** [comment]
Would be great if you could also provide an [AppImage](http://appimage.org/) for Linux. electron-builder can do this "for free".

**Comment by lgeiger** [comment]
Oh I thought we uploaded one. I can do this right now.

/cc @jdetle 😄

**Comment by lgeiger** [comment]
@probonopd I uploaded a AppImage to the 0.0.14 release. Please report if anything is broken.

**Pull request merged**